irc/server.go

```diff
@@ -188,6 +188,17 @@ func (s *Server) HandleConnectBurst(selfUser User) {
 	s.initOnce.Do(func() { close(s.initChan) })
 }
 
+// HandleChannelJoined handles a Slack-initiated channel membership change event.
+// This special signalling is necessary due to the extra data (NAMES/topic) which
+// needs to be sent by the IRC server on join.
+func (s *Server) HandleChannelJoined(channelName string) {
+	s.RLock()
+	for _, v := range s.clientConnections {
+		v.handleChannelJoined(channelName)
```

Comment:
Let's either gather the client connections inside the RLock, then run `handleChannelJoined`, or if we want atomicity run it in `s.Lock()` instead.

Reply:
I'm concerned that a client disconnect at the right time could cause us to send to an outgoingMessages channel that's no longer being polled and cause a hang on the main thread, but actually changing the locking strategy here doesn't solve that...

Reply:
Normally the goroutine in charge of sending additional `JOIN` data is the IRC client's incoming message goroutine...but in this case the additional `JOIN` data is being triggered by a Slack->IRC-server->client event. The proper behavior here would probably be to have the IRC client's outgoing message goroutine handle the query/send of additional `JOIN` data...but then we'd have to introspect the outgoing messages to see if we need to send additional data.